### DISCUSSION TITLE
Is there any shell completion for xmake?

### DISCUSSION THREAD

**Discussion by lchannng** [xmake-io/xmake#2124] [General]
### Is your feature request related to a problem? Please describe.

First of all, xmake is awlsome, thanks for all the work that you've done.
And is there any script or solution for shell completion, it's useful..i think.
Finally, thanks.

### Describe the solution you'd like

oh-my-zsh plugins..maybe

**Comment by xq114** [comment]
If you installed xmake with powershell/bash command, it should already be present

**Comment by waruqi** [comment]
you can also update to dev version `xmake update -s dev`, then run `xmake update --integrate` to reinstall shell completion scripts if it does not work.

**Comment by calyle** [comment]
Auto completion of `xmake` is not supported in zsh.
There're some issues in the completion script for zsh.
I have to run `source _xmake` before using auto completion.


I did some changes here:

```zsh
#compdef xmake xrepo

_xmake_zsh_complete() {
  local completions=("${(@f)$(XMAKE_SKIP_HISTORY=1 XMAKE_ROOT=y xmake lua private.utils.complete 0 "$words[@]")}")
  compadd -Q -S '' -- ${completions}
}

_xrepo_zsh_complete() {
  local completions=("${(@f)$(XMAKE_SKIP_HISTORY=1 XMAKE_ROOT=y xmake lua private.xrepo.complete 0 "$words[@]")}")
  compadd -Q -S '' -- ${completions}
}

compdef _xmake_zsh_complete xmake
compdef _xrepo_zsh_complete xrepo
```

cc @xq114 @waruqi

**Reply by waruqi** [reply]
Try this patch? but I did not test it. https://github.com/xmake-io/xmake/pull/7420

**Reply by calyle** [reply]
> Try this patch? but I did not test it.#7420

No, it has no effect. Auto completion can have different implementation, but `#compdef xmake xrepo` must be putted to the first line of the script, which declares this's a auto completion script.

**Reply by waruqi** [reply]
try it again

**Reply by calyle** [reply]
> try it again

Yes, it works now.

But I'm still having some issues, which may be an issue with the `xmake lua private...`

If I pass <kbd>Tab</kbd> after `xmake build`, I'll meet messed up codes.

```
xmake build ^[[0m^[[1;38;2;255;0;0;1merror: ^[[0mattempt to get length of a nil value (local 'values')^[[0m
```

These commands `clean, install, package, run, uninstall, format` have the same behaviour.

`xmake lua` should return the files in the current location as the `values`

**Reply by waruqi** [reply]
try it again

**Reply by calyle** [reply]
I have tested it, and it works fine.
